### ISSUE TITLE
`softmax` and `log_softmax` returns `nan` even when only 1 `Inf` is present

### ISSUE THREAD

**Issue by sytrus-in-github** [pytorch/pytorch#6864]
## Issue description

`F.softmax` should return one-hot representation when only 1 value is `Inf` and the others are all finite or `-Inf`. Instead, they return `nan` for all entries.

## Code example
```python
>>> import torch
>>> import torch.nn.functional as F
>>> F.softmax(torch.Tensor([0, float('-inf')]), -1)
tensor([ 1.0000,  0.0000])
>>> F.softmax(torch.Tensor([0, float('inf')]), -1)  # should give [0.0, 1.0]
tensor([    nan,     nan])
>>> F.log_softmax(torch.Tensor([0, float('-inf')]), -1)
tensor([ 0.0000,    -inf])
>>> F.log_softmax(torch.Tensor([0, float('inf')]), -1)
tensor([    nan,     nan])
>>> F.softmax(torch.Tensor([float('-inf'), 0, float('-inf')]), -1)
tensor([ 0.0000,  1.0000,  0.0000])
>>> F.softmax(torch.Tensor([0, float('inf'), 0]), -1)  # should give [0.0, 1.0, 0.0]
tensor([    nan,     nan,     nan])
>>> F.softmax(torch.Tensor([float('-inf'), 0, float('inf')]), -1)  # should give [0.0, 0.0, 1.0]
tensor([    nan,     nan,     nan])
```

## System Info

- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): conda install pytorch-nightly -c pytorch
- OS: Ubuntu 16.04
- PyTorch version: nightly 2018.04.20
- Python version: 3.6.3
- CUDA/cuDNN version: N/A
- GPU models and configuration: N/A

**Comment by ngimel** [comment]
>F.softmax should return one-hot representation when only 1 value is Inf and the others are all finite or -Inf.
This is true in the limit sense only, if one of the values is `inf` softmax is in \inf/\inf indeterminate form, so it's an open question what it should return. For most operations, limit answers won't be returned (e.g. if you try to compute sin(x)/x for x=0, the answer will be nan or fpe, not 1). I don't know if this particular case merits special treatment.

**Comment by sytrus-in-github** [comment]
Thank you for your reply!
Of course I understand that we are in a numerically tricky case. But since pytorch is trying to be friendly with edge cases: e.g. supporting `inf` and `-inf` for ops and sub-gradient etc. this might be a nice edge case to cover. I have no idea how hard it is to implement this or how bad the performance regression will be, though.

**Comment by apaszke** [comment]
Another edge case is when you have more than one infinity in the vector. What value should we return then?

**Comment by sytrus-in-github** [comment]
Of course I understand that this is a quite limited extension that we can do. But if the overhead for this extension is negligible, maybe it is worth it? Otherwise there are always DIY ways to tackle this to one's need. I just wish to discuss about this possibility. I fully understand that it might not worth it to get the code bloated for this edge case. If this is your decision, feel free to say so and close this issue :)

**Comment by apaszke** [comment]
I really don't think that's worth it. It would increase the cost of softmax kernels, while the utility is fairly limited. Once you can get one `inf` in your output, it seems likely that you can end up with more of them anyway, and this change won't save you from `nan`s. Additionally, the more popular `log_softmax` would actually have to return `-inf` for all finite elements in the input, which would likely lead to `nan`s a few ops later. In general I would avoid infinite numbers, because they will get you in all sorts of boundary conditions very easily. Hope this clarifies things.

**Comment by sytrus-in-github** [comment]
Thanks for the clarification! So, playing with `inf` is like playing with fire: it is not advisable, and if you are not careful you get burned :laughing:
